contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -167,6 +174,80 @@ pub fn set_chain_config(
     }
 }
 
+/// Applies invariants on the ITS message.
+///
+/// Invariants:
+/// - Token must be deployed on the chain before any transfers can be routed.
+/// - Token cannot be redeployed to the chain.
+/// - If the token is deployed without a custom minter, then the token is considered to be owned by ITS, and total token balance moved to the chain is tracked.
+/// - If the total token amount moved to a chain so far is `x`, then any transfer moving the token back from the chain and exceeding `x` will fail.
+///
+/// Invariants are updated depending on the ITS message type
+///
+/// 1. InterchainTransfer:
+///    - Decreases the token balance on the source chain, unless it's the origin chain in which case it's increased.
+///    - Increases the token balance on the destination chain, unless it's the origin chain in which case it's decreased.
+///    - If the balance underflows for either case, an error is returned.
+///
+/// 2. DeployInterchainToken:
+///    - If a custom minter is not set, then the token balance is tracked for the source/destination chain.
+///    - If the custom minter is set, then the balance is not tracked, but the deployment is recorded.
+///
+/// 3. DeployTokenManager:
+///    - Same as the custom minter being set above. ITS Hub can't know if the existing token on the destination chain has a custom minter set.
+fn apply_balance_invariant(
```

Comment:
Base on suggestion from Sammy, I'll separate this into individual invariants, e.g.
- deploy invariant: initialize token balance
- redeploy invariant: check if token isn't already deployed
- balance invariant: update balances and fail on underflow

Reply:
yes I like that, this function is way too broad at the moment